### ISSUE TITLE
how to ensure each sample is only seen once in one epoch

### ISSUE THREAD

**Issue by strongBrave** [webdataset/webdataset#487]
Hi,

I'm working with WebDataset to do test step and my goal is to have a standard training loop where each sample from the dataset is seen exactly once per epoch, similar to the behavior of a torch.utils.data.Dataset with a DataLoader. Cause I'm doing the test step, I want each sample just being inferenced once. However, since webdataset is a streaming dataset, I don't know how to do it properly and don't find relevant information in the DOCUMENT.

Can somebody helps me? This question really matters to me.

**Comment by tmbdev** [comment]
WebDataset by default iterates through each sample in each tar file exactly once per epoch. That's true even if you use shard shuffling and inline shuffling. They only time samples may get repeated is if you use repeat or resampling explicitly.

If you need shuffling by index with webdataset format files, see the companion webdataset/wids library.
